File: Sources/MoonRuntime/TracingLlm.swift
import Foundation

public final class TracingLlmClient: LlmClient, @unchecked Sendable {
    private let client: LlmClient
    private let writer: LlmTraceWriter

    public init(client: LlmClient, writer: LlmTraceWriter) {
        self.client = client
        self.writer = writer
    }

    public func complete(_ request: LlmRequest) async throws -> RuntimeValue {
        let start = Date()
        let messages = request.messages ?? buildMessages(request)
        do {
            let response = try await client.complete(request)
            try await writer.record(
                agent: request.agent,
                model: request.model,
                messages: messages,
                schema: request.schema,
                response: response,
                stormRound: request.stormRound,
                delegateChain: request.delegateChain,
                durationMs: Int(Date().timeIntervalSince(start) * 1000)
            )
            return response
        } catch {
            try? await writer.record(
                agent: request.agent,
                model: request.model,
                messages: messages,
                schema: request.schema,
                error: String(describing: error),
                stormRound: request.stormRound,
                delegateChain: request.delegateChain,
                durationMs: Int(Date().timeIntervalSince(start) * 1000)
            )
            throw error
        }
    }
}

File: legacy/packages/runtime/src/tracing-llm.ts
import type { LlmClient, LlmRequest } from "./llm-types";
import { LlmTraceWriter } from "./llm-trace";

export function wrapLlmWithTrace(client: LlmClient, writer: LlmTraceWriter): LlmClient {
  return {
    async complete(req: LlmRequest): Promise<unknown> {
      const start = Date.now();
      try {
        const response = await client.complete(req);
        await writer.record({
          agent: req.agent,
          model: req.model,
          messages: req.messages ?? [],
          schema: req.schema,
          response,
          stormRound: req.stormRound,
          delegateChain: req.delegateChain,
          durationMs: Date.now() - start,
        });
        return response;
      } catch (err) {
        await writer.record({
          agent: req.agent,
          model: req.model,
          messages: req.messages ?? [],
          schema: req.schema,
          error: err instanceof Error ? err.message : String(err),
          stormRound: req.stormRound,
          delegateChain: req.delegateChain,
          durationMs: Date.now() - start,
        });
        throw err;
      }
    },
  };
}